Improve documentation for Transformer name

I tried writing something like:

```go
// StatusCode is a type I declared with a String method.
diff := cmp.Diff(test.want, got, cmp.Transformer("StatusCode.String", StatusCode.String))
```

When I do this, I get a runtime panic of:

```
panic: invalid name: "StatusCode.String"
```

In the documentation for [`cmp.Transformer`](https://godoc.org/github.com/google/go-cmp/cmp#Transformer), the only mention of the name I see is at the very end:

> The name is a user provided label that is used as the Transform.Name in the transformation PathStep. If empty, an arbitrary name is used.

Here are the problem I see:

1. This doesn't help me understand what kind of name to use. As an example of the unanswered questions I'm left with: Is it human readable or used for programmatic keying later? Do I need to care for the above? Should I leave it blank? When does this show up?
2. I see no mention of what characters are allowed or not or why. Without digging into source code, I have no way of knowing why I got this panic.

@dsnet

--- a/cmp/options.go
+++ b/cmp/options.go
@@ -7,6 +7,7 @@ package cmp
 import (
 	"fmt"
 	"reflect"
+	"regexp"
 	"strings"
 
 	"github.com/google/go-cmp/cmp/internal/function"
@@ -205,6 +206,11 @@ func (invalid) apply(s *state, _, _ reflect.Value) {
 	panic(fmt.Sprintf("cannot handle unexported field: %#v\n%s", s.curPath, help))
 }
 
+// identRx represents a valid identifier according to the Go specification.
+const identRx = `[_\p{L}][_\p{L}\p{N}]*`
+
+var identsRx = regexp.MustCompile(`^` + identRx + `(\.` + identRx + `)*$`)
+
 // Transformer returns an Option that applies a transformation function that
 // converts values of a certain type into that of another.
 //
@@ -222,7 +228,9 @@ func (invalid) apply(s *state, _, _ reflect.Value) {
 // to prevent the transformer from being recursively applied upon itself.
 //
 // The name is a user provided label that is used as the Transform.Name in the
-// transformation PathStep. If empty, an arbitrary name is used.
+// transformation PathStep (and eventually shown in the Diff output).
+// The name must be a valid identifier or qualified identifier in Go syntax.
+// If empty, an arbitrary name is used.
 func Transformer(name string, f interface{}) Option {
 	v := reflect.ValueOf(f)
 	if !function.IsType(v.Type(), function.Transformer) || v.IsNil() {
@@ -231,7 +239,7 @@ func Transformer(name string, f interface{}) Option {
 	if name == "" {
 		name = "λ" // Lambda-symbol as place-holder for anonymous transformer
 	}
-	if !isValid(name) {
+	if !identsRx.MatchString(name) {
 		panic(fmt.Sprintf("invalid name: %q", name))
 	}
 	tr := &transformer{name: name, fnc: reflect.ValueOf(f)}

--- a/cmp/options_test.go
+++ b/cmp/options_test.go
@@ -113,10 +113,9 @@ func TestOptionPanic(t *testing.T) {
 		args:      []interface{}{"/*", func(int) bool { return true }},
 		wantPanic: "invalid name",
 	}, {
-		label:     "Transformer",
-		fnc:       Transformer,
-		args:      []interface{}{"_", func(int) bool { return true }},
-		wantPanic: "invalid name",
+		label: "Transformer",
+		fnc:   Transformer,
+		args:  []interface{}{"_", func(int) bool { return true }},
 	}, {
 		label:     "FilterPath",
 		fnc:       FilterPath,

--- a/cmp/path.go
+++ b/cmp/path.go
@@ -296,14 +296,3 @@ func isExported(id string) bool {
 	r, _ := utf8.DecodeRuneInString(id)
 	return unicode.IsUpper(r)
 }
-
-// isValid reports whether the identifier is valid.
-// Empty and underscore-only strings are not valid.
-func isValid(id string) bool {
-	ok := id != "" && id != "_"
-	for j, c := range id {
-		ok = ok && (j > 0 || !unicode.IsDigit(c))
-		ok = ok && (c == '_' || unicode.IsLetter(c) || unicode.IsDigit(c))
-	}
-	return ok
-}

